exporter/elasticsearchexporter/data_stream_router.go

```diff
@@ -180,11 +188,14 @@ func routeRecord(
 
 	// Only use receiver-based routing if dataset is not specified.
 	if !datasetExists {
-		// Receiver-based routing
-		// For example, hostmetricsreceiver (or hostmetricsreceiver.otel in the OTel output mode)
-		// for the scope name
-		// github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/internal/scraper/cpuscraper
-		if submatch := receiverRegex.FindStringSubmatch(scope.Name()); len(submatch) > 0 {
+		if selfTelemetryScopeNames[scope.Name()] {
+			// For collector self-telemetry, use a fixed dataset name
+			dataset = collectorSelfTelemetryDataStreamDataset
+		} else if submatch := receiverRegex.FindStringSubmatch(scope.Name()); len(submatch) > 0 {
```

Comment:
FindStringSubmatch is often the cause for heavy allocations. An alternative could look like this:

```
loc := receiverRegex.FindStringIndex(scope.Name())
if len(loc) == 0 {
   continue
}
receiverName := scope.Name()[loc[0]:loc[1]])
```